runner/src/runner.rs

```diff
@@ -100,23 +100,206 @@ impl Runner {
+            bin_name: bin_name.to_string(),
+        };
+        runner.set_downloaded_release(Some(downloaded_release.clone()));
+        Ok(downloaded_release)
+    }
+
+    fn uri_to_bin_path(runner: &impl RunnerExt, uri: &str) -> Result<(String, PathBuf), Error> {
+        // Remove any trailing slashes from the release URI
+        let parsed_uri = Url::parse(uri.trim_end_matches('/'))?;
+        let bin_name = parsed_uri
+            .path_segments()
+            .and_then(|segments| segments.last())
+            .and_then(|name| if name.is_empty() { None } else { Some(name) })
+            .ok_or(Error::ClientNameDerivationError)?;
+        let bin_path = runner.download_path().join(bin_name);
+        Ok((bin_name.to_string(), bin_path))
+    }
+
+    fn delete_downloaded_release(runner: &mut impl RunnerExt) -> Result<(), Error> {
+        let release = runner.downloaded_release().as_ref().ok_or(Error::NoDownloadedRelease)?;
+        log::info!("Removing old release, with path {:?}", release.path);
+
+        retry_with_log(
+            || Ok(fs::remove_file(&release.path)?),
+            "Failed to remove old release".to_string(),
+        )?;
+
+        runner.set_downloaded_release(None);
+        Ok(())
+    }
+
+    fn terminate_proc_and_wait(runner: &mut impl RunnerExt) -> Result<u32, Error> {
+        let child_proc = runner.child_proc().as_mut().ok_or(Error::NoChildProcess)?;
+
+        let _ = retry_with_log(
+            || {
+                Ok(signal::kill(
+                    Pid::from_raw(child_proc.id().try_into().map_err(|_| Error::IntegerConversionError)?),
+                    Signal::SIGTERM,
+                ))
+            },
+            "Failed to kill child process".to_string(),
+        )?;
+
+        match child_proc.wait() {
+            Ok(exit_status) => log::info!(
+                "Terminated vault process (pid: {}) with exit status {}",
+                child_proc.id(),
+                exit_status
+            ),
+            Err(error) => log::warn!("Vault process termination error: {}", error),
+        };
+        let pid = child_proc.id();
+        runner.set_child_proc(None);
+        Ok(pid)
+    }
+
+    async fn try_get_release<T: RunnerExt + StorageReader>(
+        runner: &T,
+        pending: bool,
+    ) -> Result<Option<ClientRelease>, Error> {
+        let storage_item = if pending {
+            PENDING_RELEASE_STORAGE_ITEM
+        } else {
+            CURRENT_RELEASE_STORAGE_ITEM
+        };
+
+        let storage_key = runner.compute_storage_key(PARACHAIN_MODULE.to_string(), storage_item.to_string());
+
+        retry_with_log_async(
+            || {
+                let key = &storage_key;
+                runner
+                    .read_chain_storage::<ClientRelease>(runner.parachain_rpc(), Some(key.to_string()))
+                    .into_future()
+                    .boxed()
+            },
+            "Error fetching executable".to_string(),
+        )
+        .await
```

Comment:
Worth using a different retry policy for this or nah?

Reply:
I think it's still fine, we don't expect RPC calls to take longer than one minute to resolve. Unless you want to increase the `RETRY_INTERVAL_MS` to reduce load on the node?